### PR TITLE
SmearGate + BigramHash + Int6 + SWA + U-Net Skips (1.1518 BPB)

### PR DESCRIPTION
**Pull request by integrate-your-mind**
## Summary

- **val_bpb: 1.1518** (int6 sliding window, stride=64, seed 1337)
- 11-layer GPT, 26.8M params, 15.2MB artifact (int6+zstd-22)
- Trained in 600s on 8×H100 SXM (9,906 steps at 60.6ms/step)

## Key Techniques

- **Per-row int6 quantization** (MLP + attention) + zstd-22 compression
- **3× MLP expansion** (hidden=1536) with **relu² activation**
- **SmearGate**: learned token-predecessor blending at input
- **BigramHash embedding**: 2048-bucket hash table (dim=128) for token-pair context
- **U-Net skip connections**: encoder→decoder with learned per-dimension weights
- **Muon optimizer** with WD=0.04, momentum warmup 0.92→0.99
- **SWA**: 7 snapshots every 200 steps during warmdown
- **Sliding window eval** (stride=64) as primary score
- **TTT LoRA eval** also included (1.1535 BPB)

## Eval Results (seed 1337)

| Method | val_loss | val_bpb | eval_time |
|--------|----------|---------|-----------|
| Pre-quantization | 1.9841 | 1.1751 | — |
| Int6 roundtrip | 2.0027 | 1.1861 | 1.9s |
| **Int6 sliding (stride=64)** | **1.9448** | **1.1518** | **97s** |
| Int6 TTT LoRA | 1.9476 | 1.1535 | 83s |

## Submission Checklist

- [x] Artifact < 16MB (15,202,515 bytes)
- [x] Trains in < 10 min on 8×H100 (600s)
- [x] Eval < 10 min (sliding 97s + TTT 83s = 180s)
- [x] train_gpt.py compiles (1191 lines, under 1500 limit)
- [x] README.md with technique descriptions
- [x] submission.json with metadata
- [x] Training log included

## Note on Seeds

Single seed submission. The improvement margin over the current #3 (1.1502) is modest, but the submission includes independently developed techniques (U-Net skips, seq_len=1024 tradeoff for more layers/steps) that may be of interest to the community. Additional seeds available on request.

## Differences from Existing Submissions

Developed independently from PR #162 (raahilshah). Key architectural differences:
- 11 layers (vs 9) with seq_len=1024 (vs 2048) — more layers + more steps
- U-Net skip connections with learned weights
- BigramHash 2048 buckets (vs 4096)
- SWA every 200 steps (vs 50)
- Includes TTT LoRA as alternative eval

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — SmearGate + BigramHash + Int6 + SWA + U-Net Skips (1.1518 BPB)

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)


## PR #289 — SmearGate + BigramHash + Int6 + SWA + U-Net Skips (1.1518 BPB)

**Head SHA:** a84d7fd2bb93803fefe2a11d7f5fad208ecb00b3  
**Author:** integrate-your-mind (Roman Mondello)  
**Date:** 2026-03-20

---

### Check 1: N-gram Family Bug (CLOSE trigger)

NOT triggered. The `BigramHashEmbedding.bigram_hash()` method computes:

```python
out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod
```

The hash key for position `i` mixes `t[i]` (current token) and `t[i-1]` (previous token). The **target token** (next token) is never in the hash key. This is structurally identical to BigramHash as defined in the legal reference — the bigram embeds the current token pair as additional input signal. CLEAN.

---

### Check 2: Pre-Quant TTT (CLOSE trigger)

NOT triggered. The TTT LoRA path (`eval_val_ttt_lora`) operates entirely **post-training**, on the serialized/dequantized model. It is called after the SWA average is applied, after quantization and roundtrip validation, and only for eval. The TTT optimizer is `Adam` (not AdamW), single-pass per document, and training is gated to `needs_train = any(ci < nc - 1 ...)` — i.e., only trains on non-final chunks so the final chunk score is computed clean. No multi-epoch AdamW on val_tokens without score-first. CLEAN.

---

### Check 3: Legal TTT (CLEAN)

Confirmed. The per-chunk logic in `eval_val_ttt_lora` follows score-first discipline: the final chunk of each document (`ci == num_chunks - 1`) is evaluated with `torch.no_grad()` and does not participate in backward. The training signal (`needs_train`) is explicitly masked off for final chunks, and `_accumulate_bpb` runs under `torch.no_grad()`. This matches the PR #1413 legal TTT pattern. CLEAN.

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*
